src/modules/scopes.js

```diff
@@ -72,63 +72,74 @@ function decorateRootScope($delegate, $parse) {
 
-  // var _watch = scopePrototype.$watch;
-  // scopePrototype.$watch = function (watchExpression, reactionFunction) {
-  //   var watchStr = humanReadableWatchExpression(watchExpression);
-  //   var scopeId = this.$id;
-  //   if (typeof watchExpression === 'function') {
-  //     arguments[0] = function () {
-  //       var start = perf.now();
-  //       var ret = watchExpression.apply(this, arguments);
-  //       var end = perf.now();
-  //       hint.emit('scope:watch', {
-  //         id: scopeId,
-  //         watch: watchStr,
-  //         time: end - start
-  //       });
-  //       return ret;
-  //     };
-  //   } else {
-  //     var thatScope = this;
-  //     arguments[0] = function () {
-  //       var start = perf.now();
-  //       var ret = thatScope.$eval(watchExpression);
-  //       var end = perf.now();
-  //       hint.emit('scope:watch', {
-  //         id: scopeId,
-  //         watch: watchStr,
-  //         time: end - start
-  //       });
-  //       return ret;
-  //     };
-  //   }
-
-  //   if (typeof reactionFunction === 'function') {
-  //     var applyStr = reactionFunction.toString();
-  //     arguments[1] = function () {
-  //       var start = perf.now();
-  //       var ret = reactionFunction.apply(this, arguments);
-  //       var end = perf.now();
-  //       hint.emit('scope:reaction', {
-  //         id: this.$id,
-  //         watch: watchStr,
-  //         time: end - start
-  //       });
-  //       return ret;
-  //     };
-  //   }
-
-  //   return _watch.apply(this, arguments);
-  // };
+  var _watch = scopePrototype.$watch;
+  var _digestEvents = [];
+  scopePrototype.$watch = function (watchExpression, reactionFunction) {
+    var watchStr = humanReadableWatchExpression(watchExpression);
+    var scopeId = this.$id;
+    if (typeof watchExpression === 'function') {
+      arguments[0] = function () {
+        var start = perf.now();
+        var ret = watchExpression.apply(this, arguments);
+        var end = perf.now();
+        _digestEvents.push({
+          eventType: 'scope:watch',
+          id: scopeId,
+          watch: watchStr,
+          time: end - start
+        });
+        return ret;
+      };
+    } else {
+      var thatScope = this;
+      arguments[0] = function () {
+        var start = perf.now();
+        var ret = thatScope.$eval(watchExpression);
+        var end = perf.now();
+        _digestEvents.push({
+          eventType: 'scope:watch',
+          id: scopeId,
+          watch: watchStr,
+          time: end - start
+        });
+        return ret;
+      };
+    }
+
+    if (typeof reactionFunction === 'function') {
+      arguments[1] = function () {
+        var start = perf.now();
+        var ret = reactionFunction.apply(this, arguments);
+        var end = perf.now();
+        _digestEvents.push({
+          eventType: 'scope:reaction',
+          id: this.$id,
+          watch: watchStr,
+          time: end - start
+        });
+        return ret;
```

Comment:
this function and the previous functions before it that are getting assigned to arguments seem to only have very small differences (the ret and the eventType?). would it be possible to refactor it and move them into a function or maybe create a decorator? let me know if this isn't doable. (still trying to get a feel for the code etc)